transformer_engine/pytorch/float8_tensor.py

```diff
@@ -435,80 +463,50 @@ def expand_as(self, other: torch.Tensor):
-
+        cache: bool, default = `False`
+               Whether or not to cache the transpose.
+        noop_flag: Optional[torch.Tensor], default = `None`
+                   Only used if argument `cache` is `True`, ignored otherwise.
+                   A single element fp32 tensor with a value of 1.0 or 0.0
+                   which is treated as a boolean. `1.0` forces recompute
+                   and `0.0` executes a noop using the same kernel.
         """
+        assert self.dim() == 2, f"{self.dim()}-D transpose not supported."
 
-        # Check caching mode
-        if not isinstance(update_cache, str):
-            update_cache = "force" if update_cache else "reuse_only"
-        if update_cache not in ("force", "reuse_only", "lazy"):
-            raise ValueError(
-                "Supported values for update_cache are "
-                '"force" (True), "reuse_only" (False), "lazy" '
-                f"(got {update_cache})"
-            )
+        # Case: no caching.
+        if not cache:
+            return tex.fp8_transpose(self._data, self._fp8_dtype)
 
-        # Handle non-2D transposes
-        if -self.dim() <= dim0 < 0:
-            dim0 += self.dim()
-        if -self.dim() <= dim1 < 0:
-            dim1 += self.dim()
-        if self.dim() != 2 or dim0 == dim1:
-            if update_cache == "force":
-                raise ValueError(
-                    "Transpose caching is only supported for basic 2D transposes "
-                    f"(ndims={self.dim()}, dim0={dim0}, dim1={dim1})"
-                )
-            return super().transpose(dim0, dim1)
-
-        # Clear cache if needed
-        if update_cache == "force":
-            self._transpose = None
-
-        # Compute transpose if needed
-        out = self._transpose
-        if out is None:
-            out = Float8Tensor.make_like(
-                self,
-                data=tex.fp8_transpose(
-                    self._data.contiguous(),
-                    self._fp8_dtype,
-                ),
-            )
+        # Case: reuse cache without calling a kernel.
+        if not self._transpose_invalid and noop_flag is None:
+            assert self._transpose is not None, "Tranpose cache is empty."
+            return self._transpose
 
-        # Update cache if needed
-        if update_cache in ("force", "lazy"):
-            self._transpose = out
-        return out
+        # Allocate transpose if needed.
+        if self._transpose is None:
+            shape = (self._data.shape[1], self._data.shape[0])
+            self._transpose = torch.empty(shape, dtype=torch.uint8, device=self._data.device)
+
+        # Case: recompute transpose and store cache.
+        if noop_flag is None:
+            tex.fp8_transpose_noalloc(self._data, self._transpose, self._fp8_dtype)
+        else:
+            # Case: cuda graph capture.
+            tex.fp8_transpose_noalloc_noop(self._data, self._transpose, noop_flag, self._fp8_dtype)
```

Comment:
In #707, I found it useful to implicitly flatten tensors to 2D:
```suggestion
        # Allocate transpose if needed.
        data_2d = self._data.reshape(-1, self._data.shape[-1])
        if self._transpose is None:
            shape = (self._data.shape[-1], math.prod(self._data.shape[:-1]))
            self._transpose = torch.empty(shape, dtype=torch.uint8, device=self._data.device)

        # Case: recompute transpose and store cache.
        if noop_flag is None:
            tex.fp8_transpose_noalloc(data_2d, self._transpose, self._fp8_dtype)
        else:
            # Case: cuda graph capture.
            tex.fp8_transpose_noalloc_noop(data_2d, self._transpose, noop_flag, self._fp8_dtype)
```
This is because most layers do their compute in 2D and reshapes to BSHD/SBHD. The next layer will then just reshape it back to 2D, so it would be nice for the transpose cache to survive that reshaping. See this helper function for reshaping in #707: https://github.com/NVIDIA/TransformerEngine/blob/1188d9e459fa0a7aeb130219ec94c9ba27356735/transformer_engine/pytorch/fuser/ops/_common.py#L132